src/cluster/OnLoop.h

```diff
@@ -139,43 +157,58 @@ class OnLoopProcess : public zeek::iosource::IOSource {
         bool fire = false;
         size_t qs = 0;
 
-        while ( ! to_queue.empty() ) {
-            {
-                std::scoped_lock lock(mtx);
-
-                if ( ! IsOpen() ) {
-                    // IO Source is being removed.
-                    fire = false;
-                    break;
-                }
-
-                qs = queue.size();
-                if ( qs < max_queue_size ) {
-                    queue.splice(queue.end(), to_queue);
-                    fire = fire || qs == 0;
-                    assert(to_queue.empty());
-                    assert(! queue.empty());
-                }
+        int timeouts = 0;
+
+        {
+            std::unique_lock lock(mtx);
+
+            // Wait for room in the queue.
+            while ( IsOpen() && queue.size() >= max_queue_size ) {
+                auto status = cond.wait_for(lock, cond_timeout);
+                if ( status == std::cv_status::timeout && IsOpen() )
+                    ++timeouts;
             }
 
-            if ( ! to_queue.empty() ) {
-                std::this_thread::sleep_for(block_duration);
-                fire = true;
+            if ( IsOpen() ) {
+                assert(queue.size() < max_queue_size);
+                assert(to_queue.size() == 1);
+                queue.splice(queue.end(), to_queue);
+                fire = queue.size() == 1; // first element in queue triggers processing.
+            }
+            else {
+                // IO Source is being or was removed.
+                fire = false;
             }
         }
 
+
         if ( fire )
             flare.Fire();
 
+        if ( timeouts > 0 ) {
+            // XXX: Should this invoke some callback or change the return value
+            // so users can react on this?
+            //
+            // We could also do suicidal snail pattern here. If the event
+            // loop is unable to process, we may as well knock ourselves out.
+            std::fprintf(stderr, "timeouts %d!\n", timeouts);
```

Comment:
Like discussed offline, this probably should just be a counter in a metric somewhere.

Reply:
Done.